### PR TITLE
<format>: Fix arg overload resolution

### PR DESCRIPTION
**Pull request by eldakesh-ms**
There seems to be a compiler bug picking the wrong function in an
overload set. I've filed it as DevCom-1390432. This sidesteps the bug, and also
fixes a latent bug with `_Get_format_arg_storage_type`.

The issue was that the overload for `const _Ty&` was being picked for
`_Store`, so a `void*` was stored as a handle, but the `_Get_format_arg_storage_type`
overload that was being picked was the one for void pointers. This cause us to
buffer overrun the storage. Now it always picks the `const _Ty&` overload
(which seems weird but that is what the overload resolution is supposed to be).

There is also a fix for the packed index, that wasn't very packed
because the types of the bitfield were different. This should save space
as intended.

### CODE REVIEW

File: tests/std/tests/P0645R10_text_formatting_args/test.cpp [hunk 1]
```diff
@@ -213,6 +213,8 @@ void test_format_arg_store() {
     test_single_format_arg<Context, basic_string_view<char_type>, Arg_type::string_type>(get_input_sv<char_type>());
 }
```
Comment (CaseyCarter) [inline tests/std/tests/P0645R10_text_formatting_args/test.cpp:215]: Missing regression test for the other bug we're fixing.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/format [hunk 2]
```diff
@@ -976,19 +976,21 @@ template <class _Context, class _Ty>
...
+    return static_cast<const typename _Context::char_type*>(nullptr);
 }
 
-template <class _Context, class _Traits, class _Char_type = typename _Context::char_type>
-/* consteval */ constexpr auto _Get_format_arg_storage_type(basic_string_view<_Char_type, _Traits>) noexcept {
-    return basic_string_view<_Char_type>{};
+template <class _Context, class _Traits>
+/* consteval */ constexpr auto _Get_format_arg_storage_type(
+    basic_string_view<typename _Context::char_type, _Traits>) noexcept {
+    return basic_string_view<typename _Context::char_type>{};
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/format:987]: Answering my own question: because the working draft says so. We ignore (instead of reject) non-default traits and simply allow the initialization of the exposition-only variant to be ill-formed and explode. That will produce _horrible_ error messages.

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/format:987]: Clarifying: the code already forces said conversion, but to my knowledge nobody has filed an LWG issue to allow that behavior.

File: stl/inc/format [hunk 3]
```diff
@@ -1006,18 +1008,19 @@ template <class _Context, class _Ty>
...
 
     constexpr _Format_arg_store_packed_index() = default;
     constexpr explicit _Format_arg_store_packed_index(const size_t _Index)
-        : _Index(static_cast<_Index_type>(_Index)), _Type(_Basic_format_arg_type::_None) {}
+        : _Index(static_cast<_Index_type>(_Index)), _Type(static_cast<_Index_type>(_Basic_format_arg_type::_None)) {}
+
 
     _Index_type _Index : (sizeof(_Index_type) * 8 - 4);
-    _Basic_format_arg_type _Type : 4;
+    _Index_type _Type : 4;
```
Comment (miscco) [inline stl/inc/format:1023]: What was wrong here? I thought it should have packed those

Comment (eldakesh-ms) [reply stl/inc/format:1023]: I think the general rule MSVC uses for bitfields is unrelated types don't pack well together, and it seems only equal size primitives pack together.

File: stl/inc/format [hunk 4]
```diff
@@ -1006,18 +1008,19 @@ template <class _Context, class _Ty>
 
 template <class _Context, class _Ty>
 inline constexpr size_t _Get_format_arg_storage_size = sizeof(
-    _Get_format_arg_storage_type<_Context>(_STD declval<_Ty>()));
+    _Get_format_arg_storage_type<_Context>(static_cast<const _Ty&>(_STD declval<_Ty>())));
```
[the remark this answers is not in the document]
Comment (eldakesh-ms) [reply stl/inc/format:1011]: So if I understand correctly, the issue is basically that a users construction of an rvalue `void*` `basic_format_arg` would result in a storage of a `_Pointer_type` but being handled through `make_format_args` makes a handle? Is that really an issue? The way that `basic_format_arg` is currently written, this compiles fine: `basic_format_arg<format_context>(static_cast<void*>(nullptr))`.

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/format:1011]: My counterexample was a type with a user-defined conversion to `float` but only from non-`const` `meow`. I now believe it to be bogus since there's no way for a user to directly construct a `basic_format_args` - only the library can do so, and it does so only from functions that take `const Args&...` so they've already been `const`-ed when they get to this overload set.

TLDR: this comment is now simply:
```suggestion
    _Get_format_arg_storage_type<_Context>(_STD declval<const _Ty&>()));
```

File: stl/inc/format [hunk 5]
```diff
@@ -1045,7 +1048,7 @@ private:
         const auto _Length      = _Get_format_arg_type_storage_size<_CharType>(_Arg_type);
 
         _CSTD memcpy(_Storage + _Index_length + _Store_index, _STD addressof(_Val), _Length);
-        _Index_array[_Arg_index]._Type = _Arg_type;
+        _Index_array[_Arg_index]._Type = static_cast<_Index_type::_Index_type>(_Arg_type);
```
Comment (barcharcraz) [inline stl/inc/format:1051]: can we just attach a member function to _Index_type that does this?

File: stl/inc/format [hunk 6]
```diff
@@ -1151,7 +1154,7 @@ public:
         const auto _Index_length          = _Num_args * sizeof(_Format_arg_store_packed_index);
         const unsigned char* _Arg_storage = _Storage + _Index_length + _Packed_index._Index;
 
-        switch (_Packed_index._Type) {
+        switch (static_cast<_Basic_format_arg_type>(_Packed_index._Type)) {
```
Comment (barcharcraz) [inline stl/inc/format:1157]: similarly can we attach a member function to _Index_type that does this.

**Review by barcharcraz — CHANGES_REQUESTED** [review]

File: stl/inc/format (hunk 5 above)
Comment (eldakesh-ms) [reply stl/inc/format:1051]: I originally did that, but it's only done once to assign and once to access, so it felt kind of needlessly wordy. I still can, I just wasn't a fan of it.

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/format (hunk 2 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/format:987]: I'd be fine with adding a card to the Format Project as a reminder to file an LWG issue; shouldn't block merging this.

**Review by barcharcraz — APPROVED** [review]

File: stl/inc/format (hunk 5 above)
Comment (miscco) [reply stl/inc/format]: I would rather not. If it is rare, we can avoid the debug hit

File: stl/inc/format (hunk 2 above)
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/format:987]: I've recorded this in my list of issues; I'll file these today either here or with LWG.
